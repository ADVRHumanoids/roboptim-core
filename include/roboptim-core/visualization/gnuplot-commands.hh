// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Gnuplot visualization.
 */

#ifndef ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_COMMANDS_HH
# define ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_COMMANDS_HH
# include <boost/format.hpp>

namespace roboptim
{
  namespace visualization
  {
    namespace gnuplot
    {
      class Command
      {
      public:
	explicit Command (std::string cmd) throw ()
	  : command_ (cmd)
	{}

	~Command () throw () {}

	const std::string& command () const throw ()
	{
	  return command_;
	}
      protected:
	std::string command_;
      };

# define GNUPLOT_UNARY_COMMAND(NAME)		\
      Command					\
      NAME () throw ()				\
      {						\
	return Command (#NAME);			\
      }

# define GNUPLOT_STR_COMMAND(NAME, FARG, ARG)	\
      Command					\
      NAME (const char* FARG) throw ()		\
      {						\
	std::string command = #NAME;		\
	if (*ARG != 0)				\
	  {					\
	    command += " '";			\
	    command += ARG;			\
	    command += "'";			\
	  }					\
	return Command (command);		\
      }

      Command
      comment (const char* content) throw ()
      {
	std::string str = "#";
	str += content;
	return Command (str);
      }

      Command
      set (const char* var, const char* value) throw ()
      {
	return Command ((boost::format ("set %1% %2%") % var % value).str ());
      }

      Command
      show (const char* var, const char* value) throw ()
      {
	return Command ((boost::format ("show %1% %2%") % var % value).str ());
      }

      GNUPLOT_UNARY_COMMAND (clear)
      GNUPLOT_UNARY_COMMAND (pwd)
      GNUPLOT_UNARY_COMMAND (quit)
      GNUPLOT_UNARY_COMMAND (replot)
      GNUPLOT_UNARY_COMMAND (reread)
      GNUPLOT_UNARY_COMMAND (reset)

      GNUPLOT_STR_COMMAND(cd, path, path)
      GNUPLOT_STR_COMMAND(help, topic = "", topic)


# undef GNUPLOT_STR_COMMAND
# undef GNUPLOT_UNARY_COMMAND

    } // end of namespace gnuplot.

  } // end of namespace visualization.
} // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_COMMANDS_HH
