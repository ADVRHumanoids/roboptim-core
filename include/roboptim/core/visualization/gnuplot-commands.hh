// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_COMMANDS_HH
# define ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_COMMANDS_HH
# include <roboptim/core/sys.hh>
# include <roboptim/core/debug.hh>

# include <string>

namespace roboptim
{
  namespace visualization
  {
    namespace gnuplot
    {
      /// \addtogroup roboptim_visualization
      /// @{

      /// \brief Gnuplot command.
      ///
      /// A command represents a Gnuplot command.
      /// The representation of the command as a string
      /// can be accessed through the #command() method.
      class Command
      {
      public:
	/// \brief Make a command from a string.
	explicit Command (std::string cmd) throw ();

	~Command () throw ();

	/// \brief Retrieve the command as a string.
	const std::string& command () const throw ();
      protected:
	/// \brief Store Gnuplot command.
	std::string command_;
      };

      /// \brief Make a Gnuplot comment.
      Command comment (const char*) throw ();


      /// \brief Make a Gnuplot set command.
      ///
      /// Make a Gnuplot set command (assign a new
      /// value to a variable in Gnuplot).
      /// \param var variable that will be set
      /// \param val new variable value
      /// \return Gnuplot command
      Command set (const char* var, const char* val = "") throw ();

      /// \brief Make a Gnuplot unset command.
      ///
      /// Make a Gnuplot unset command (destroy a variable).
      /// \param var variable that will be unset
      /// \return Gnuplot command
      Command unset (const char* var) throw ();

      /// \brief Make a Gnuplot show command.
      ///
      /// Make a Gnuplot show command (show a variable value).
      /// \param var variable that will be shown
      /// \return Gnuplot command
      Command show (const char* var) throw ();

      /// \brief Make a Gnuplot clear command.
      ///
      /// Make a Gnuplot clear command (clear the screen).
      /// \return Gnuplot command
      Command clear () throw ();

      /// \brief Make a Gnuplot pwdcommand.
      ///
      /// Make a Gnuplot pwd command (show the current directory).
      /// \return Gnuplot command
      Command pwd () throw ();

      /// \brief Make a Gnuplot quit command.
      ///
      /// Make a Gnuplot quit command (quit Gnuplot).
      /// \return Gnuplot command
      Command quit () throw ();

      /// \brief Make a Gnuplot replot command.
      ///
      /// Make a Gnuplot replot command (redraw previously defined curves).
      /// \return Gnuplot command
      Command replot () throw ();

      /// \brief Make a Gnuplot reread command.
      ///
      /// Make a Gnuplot reread command (reread data files).
      /// \return Gnuplot command
      Command reread () throw ();

      /// \brief Make a Gnuplot reset command.
      ///
      /// Make a Gnuplot replot command (reset Gnuplot state).
      /// \return Gnuplot command
      Command reset () throw ();

      /// \brief Make a Gnuplot cd command.
      ///
      /// Make a Gnuplot cd command (change current directory).
      /// \param dir new directory
      /// \return Gnuplot command
      Command cd (const char* dir) throw ();

      /// \brief Make a Gnuplot help command.
      ///
      /// Make a Gnuplot help command (display Gnuplot help).
      /// \param topic help topic
      /// \return Gnuplot command
      Command help (const char* topic = "") throw ();

      /// @}

    } // end of namespace gnuplot.
  } // end of namespace visualization.
} // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_COMMANDS_HH
