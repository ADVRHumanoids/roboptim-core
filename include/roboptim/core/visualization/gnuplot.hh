// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Gnuplot visualization.
 */

#ifndef ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_HH
# define ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_HH
# include <vector>

# include <roboptim/core/visualization/fwd.hh>
# include <roboptim/core/visualization/gnuplot-commands.hh>

namespace roboptim
{
  namespace visualization
  {
    class Gnuplot
    {
    public:
      ~Gnuplot () throw ();

      /// Instanciate a Gnuplot without setting a term.
      static Gnuplot make_gnuplot () throw ()
      {
	return Gnuplot ();
      }

      /// Instanciate a Gnuplot suitable for interactive use.
      /// (i.e. being seen by an user, not just rendering an image).
      static Gnuplot make_interactive_gnuplot () throw ()
      {
	Gnuplot gp;
	gp.push_command (gnuplot::set ("terminal", "wxt persist"));
	return gp;
      }

      void push_command (gnuplot::Command) throw ();

      std::ostream& print (std::ostream&) const throw ();

      Gnuplot& operator << (gnuplot::Command) throw ();

    protected:
      explicit Gnuplot () throw ();
    private:
      std::vector<gnuplot::Command> commands_;
    };

    template <typename T>
    Gnuplot& operator<< (Gnuplot& gp, T t)
    {
      gp.push_command (t ());
      return gp;
    }

    std::ostream& operator<< (std::ostream&, const Gnuplot&);
  } // end of namespace visualization.
} // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_HH
