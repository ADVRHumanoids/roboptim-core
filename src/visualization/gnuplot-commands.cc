// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <boost/format.hpp>

#include <roboptim/core/visualization/gnuplot-commands.hh>

namespace roboptim
{
  namespace visualization
  {
    namespace gnuplot
    {
      Command::Command (std::string cmd) throw ()
	: command_ (cmd)
      {}

      Command::~Command () throw ()
      {}

      const std::string&
      Command::command () const throw ()
      {
	return command_;
      }


# define GNUPLOT_UNARY_COMMAND(NAME)		\
      Command					\
      NAME () throw ()				\
      {						\
	return Command (#NAME);			\
      }

# define GNUPLOT_STR_COMMAND(NAME, ARG)		\
      Command					\
      NAME (const char* ARG) throw ()		\
      {						\
	std::string command = #NAME;		\
	if (*ARG != 0)				\
	  {					\
	    command += " '";			\
	    command += ARG;			\
	    command += "'";			\
	  }					\
	return Command (command);		\
      }

      Command
      comment (const char* content) throw ()
      {
	std::string str = "#";
	str += content;
	return Command (str);
      }

      Command
      set (const char* var, const char* value) throw ()
      {
	if (!*value)
	  return Command ((boost::format ("set %1%") % var).str ());
	return Command ((boost::format ("set %1% %2%") % var % value).str ());
      }

      Command
      unset (const char* var) throw ()
      {
	return Command ((boost::format ("unset %1%") % var).str ());
      }

      Command
      show (const char* var) throw ()
      {
	return Command ((boost::format ("show %1%") % var).str ());
      }

      GNUPLOT_UNARY_COMMAND (clear)
      GNUPLOT_UNARY_COMMAND (pwd)
      GNUPLOT_UNARY_COMMAND (quit)
      GNUPLOT_UNARY_COMMAND (replot)
      GNUPLOT_UNARY_COMMAND (reread)
      GNUPLOT_UNARY_COMMAND (reset)

      GNUPLOT_STR_COMMAND (cd, path)
      GNUPLOT_STR_COMMAND (help, topic)

# undef GNUPLOT_STR_COMMAND
# undef GNUPLOT_UNARY_COMMAND

    } // end of namespace gnuplot.
  } // end of namespace visualization.
} // end of namespace roboptim
