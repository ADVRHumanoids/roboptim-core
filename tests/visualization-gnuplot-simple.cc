// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/fixture.hh"

#include <iostream>

#include <roboptim/core/io.hh>

#include <roboptim/core/visualization/gnuplot.hh>
#include <roboptim/core/visualization/gnuplot-commands.hh>

using namespace roboptim;
using namespace roboptim::visualization;

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE (visualization_gnuplot_simple)
{
  boost::shared_ptr<boost::test_tools::output_test_stream>
    output = retrievePattern ("visualization-gnuplot-simple");

  using namespace roboptim::visualization::gnuplot;
  Gnuplot gnuplot = Gnuplot::make_gnuplot ();

  // Display nothing, generate just some valid Gnuplot commands.
  (*output)
    << (gnuplot
	<< set ("term", "x11 enhanced persist")
	<< comment ("Hello, world!")
	<< cd ("/tmp")
	<< help ("pwd")
	<< show ("foo")
	<< replot
	<< pwd
	<< reread
	<< clear
	<< reset
	<< quit
	);

  // Test clear function
  gnuplot.clear ();
  (*output) << gnuplot;

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}

BOOST_AUTO_TEST_SUITE_END ()
