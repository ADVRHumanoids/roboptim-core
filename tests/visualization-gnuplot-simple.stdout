#!/usr/bin/gnuplot
# Generated by RobOptim core library 0.1
set term x11 enhanced persist
#Hello, world!
cd '/tmp'
reset
